compiler/src/dotty/tools/dotc/core/Scopes.scala

```diff
@@ -39,6 +39,12 @@ object Scopes {
    */
   private final val MaxRecursions = 1000
 
+  /** A function that optionally produces synthesized symbols with
+   *  the given name in the given context. Returns `NoSymbol` if the
+   *  no symbol should be synthesized for the given name.
+   */
+  type SymbolSynthesizer = Name => Context => Symbol
```

Comment:
I would assume that with our current simple use cases the JVM must be capable of inlineing the this additional lambda. Seem a bit premature optimization.